kimchi/src/circuits/polynomials/not.rs

```diff
@@ -0,0 +1,235 @@
+    // - double  : whether to perform two NOTs or only one inside the generic gate
+    // Output:
+    // - new_row : next row after the double NOT generic gate, corresponds to `new_row+1`
+    fn not_gnrc(gates: &mut Vec<Self>, pub_row: usize, new_row: usize, double: bool) -> usize {
+        let g1 = GenericGateSpec::Add {
+            left_coeff: None,
+            right_coeff: Some(-F::one()),
+            output_coeff: None,
+        };
+        let g2 = {
+            if double {
+                Some(GenericGateSpec::Add {
+                    left_coeff: None,
+                    right_coeff: Some(-F::one()),
+                    output_coeff: None,
+                })
+            } else {
+                None
+            }
+        };
+        let mut not_gate = vec![CircuitGate::create_generic_gadget(
+            Wire::for_row(new_row),
+            g1,
+            g2,
+        )];
+        gates.append(&mut not_gate);
+        // check left inputs of the double generic gate correspond to the 2^bits - 1 value
+        gates.connect_cell_pair((pub_row, 0), (new_row, 0));
+        if double {
+            gates.connect_cell_pair((pub_row, 0), (new_row, 3));
+        }
+        new_row + 1
+    }
+
+    /// Creates a NOT gadget for `bits` length using Xor gates.
+    /// It implicitly constrains the length of the input to be at most 16 * num_xors bits.
+    /// Includes:
+    /// - num_xors Xor16 gates
+    /// - 1 Generic gate to constrain the final row to be zero with itself
+    /// Input:
+    /// - new_row : row to start the NOT gadget
+    /// Requires:
+    /// - 1 initial public input generic gate in `pub_row` to constrain the input to be `2^bits-1`.
+    /// INTEGRATION:
+    /// - Connect the left input to a public input row containing the 2^bits-1 value
+    pub fn extend_not_xor_gadget(
+        gates: &mut Vec<Self>,
+        pub_row: usize,
+        new_row: usize,
+        bits: usize,
+    ) -> usize {
+        let num_xors = num_xors(bits);
+        let mut not_gates = (0..num_xors)
+            .map(|i| CircuitGate {
+                typ: GateType::Xor16,
+                wires: Wire::for_row(new_row + i),
+                coeffs: vec![],
+            })
+            .collect::<Vec<_>>();
+        let zero_row = new_row + num_xors;
+        not_gates.push(CircuitGate::create_generic_gadget(
+            Wire::for_row(zero_row),
+            GenericGateSpec::Const(F::zero()),
+            None,
+        ));
+        gates.extend(not_gates);
+        // check fin_in1, fin_in2, fin_out are zero
+        gates.connect_cell_pair((zero_row, 0), (zero_row, 1));
+        gates.connect_cell_pair((zero_row, 0), (zero_row, 2));
+        // Integration
+        gates.connect_cell_pair((pub_row, 0), (new_row, 1)); // input2 of xor is all ones
+
+        gates.len()
+    }
+}
+
+/// Extend a Not witness for less than 255 bits (native field)
+/// Input: full witness, first input and optional bit length
```

Comment:
If `bits` is set it gets the max. Meaning, if I set for example bits to be 3 and I try to negate word `0b001` then it returns `0b110` instead of `0b0`, which would have been the output if bits was `None`.

Reply:
Ah, ok - makes sense now.